src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs

```diff
@@ -50,36 +53,101 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
+                // Always keep track of its step information, regardless of if the entry was removed or not, so we
+                // can accurately report how long it took and what actually happened (for testing validation).
                 sourceInputsBuilder?.Add((entry.Step!, entry.OutputIndex));
+
+                if (entry.State != EntryState.Removed)
+                    entryCount++;
             }
 
-            var sourceValues = sourceValuesBuilder.ToImmutableAndFree();
-            var sourceInputs = newTable.TrackIncrementalSteps ? sourceInputsBuilder!.ToImmutableAndFree() : default;
+            var sourceInputs = sourceInputsBuilder != null ? sourceInputsBuilder.ToImmutableAndFree() : default;
 
-            if (previousTable.IsEmpty)
+            // First, see if we can reuse the entries from the previous table.
+            var result = tryReusePreviousTableValues(entryCount);
+            if (result.IsDefault)
             {
-                newTable.AddEntry(sourceValues, EntryState.Added, stopwatch.Elapsed, sourceInputs, EntryState.Added);
+                // If not, produce the actual values we need from sourceTable.
+                result = computeCurrentTableValues(entryCount);
             }
-            else if (!sourceTable.IsCached || !newTable.TryUseCachedEntries(stopwatch.Elapsed, sourceInputs))
+
+            return (result, sourceInputs);
+
+            ImmutableArray<TInput> tryReusePreviousTableValues(int entryCount)
             {
-                if (!newTable.TryModifyEntry(sourceValues, _comparer, stopwatch.Elapsed, sourceInputs, EntryState.Modified))
+                if (previousTable.Count != 1)
+                    return default;
+
+                var previousItems = previousTable.Single().item;
+
+                // If they don't have the same length, we clearly can't reuse them.
+                if (previousItems.Length != entryCount)
+                    return default;
+
+                var indexInPrevious = 0;
+                foreach (var entry in sourceTable)
                 {
-                    newTable.AddEntry(sourceValues, EntryState.Added, stopwatch.Elapsed, sourceInputs, EntryState.Added);
+                    if (entry.State == EntryState.Removed)
+                        continue;
+
+                    // If the entries aren't the same, we can't reuse.
+                    if (!EqualityComparer<TInput>.Default.Equals(entry.Item, previousItems[indexInPrevious]))
+                        return default;
+
+                    indexInPrevious++;
                 }
+
+                // We better have the exact same count as previousItems as we checked that above.
+                Debug.Assert(indexInPrevious == previousItems.Length);
+
+                // Looks good, we can reuse this.
+                return previousItems;
             }
 
-            return newTable.ToImmutableAndFree();
+            ImmutableArray<TInput> computeCurrentTableValues(int entryCount)
```

Comment:
finally, we do a pass where we allocate a builder.  however, importantly, we know the exact amount of items we need, so the builder doesn't create a scratch array that has the chance of not being pooled if too large.